### ISSUE TITLE
lineardb3 Databases fail when filesize exceeds 2GB

### ISSUE THREAD

**Issue by risvh** [jasonrohrer/OneLife#838]
**TLDR:**

When map.db or mapTime.db exceeds 2GB (more precisely, the value of max signed int, in bytes), they stop working. That causes the game to glitch out in various ways. For example, in the former case, objects may disappear when dropped; or in the latter case, more subtly, trees planted at the same time grow out of sync (more so allowed by the intended random variation), some of the plots may even stuck ungrown. This bug has happened in both OHOL and 2HOL.

**Long story:**

For quite some time now, 2HOL maps couldn't last longer than about one year when maps start to glitch out in the ways mentioned above. Only after several maps we observed the pattern about the filesize of the databases.

![](https://github.com/jasonrohrer/OneLife/assets/67486979/64c16ffe-6904-4f53-b4fe-aad74333aee8)

Here is a video when map.db exceeds 2GB on 2HOL around Sep 2022:

https://github.com/jasonrohrer/OneLife/assets/67486979/e562cf90-e38f-41b2-9006-befd2e475830

Another video when mapTime.db exceeds 2GB in May 2023. It is normal that the character has to move near the plots for the transition to happen. The weird thing here is that some plots simply refuse to transition to the next growth stage:

https://github.com/jasonrohrer/OneLife/assets/67486979/9c9f105b-b739-44e5-8943-36d2b06fa3e4

And finally this, the Time Bug, happened on OHOL right before April Fool update in 2023. The server then was up without restarts for a long period of time. (Server restarts in OHOL slightly relieve the issue, more on this below.)

![](https://github.com/jasonrohrer/OneLife/assets/67486979/b20c95d5-0262-42c9-8150-1fd81300c3d6)

Gonna discuss these here for the record, may not be 100% accurate but this is my understanding of how these work. There are 2 mechanics in the game for "civilization being reclaimed by nature".

First one is Culling, relevant setting file [longTermNoLookCullSeconds.ini](https://github.com/jasonrohrer/OneLife/blob/master/server/settings/longTermNoLookCullSeconds.ini). When a cell is "not looked" for some period of time, the cell gets "culled", meaning it reverts to its natural/wild state from the map generation. This simply means that the cell is overwritten in the database with whatever the natural value is, NOT removing the cell record from the database (Deletion is not a thing/not supported in lineardb3). Database sizes are NOT reduced by this mechanics. This piece of mechanics takes place periodically when the server is running.

The second one is LookTimeCleanup, relevant setting file [mapCellForgottenSeconds.ini](https://github.com/jasonrohrer/OneLife/blob/master/server/settings/mapCellForgottenSeconds.ini). When the server starts up, it creates temporary databases and selectively copy records from the old ones, leaving out the cells where the last lookTime is too long ago, hence really shrinking the database sizes. This is why the Time Bug went away on OHOL right around the April Fool when the server was restarted for the update.

From the experience on both 2HOL and OHOL, database sizes reach the limit of 2GB roughly one year or so of continuous server up time.

I am not sure where in the database architecture this max signed int limitation arises, or if this limit is intentional / you are aware of this limit (related to the game being 32-bit?). I am leaving this issue here in hope that you could provide more information about it. Thank you.

**Comment by jasonrohrer** [comment]
Wow, great find!

I actually heard some other reports of strange behavior where one tree would grow and another would not.  So this likely explains it.

Looking into fixing it!

**Comment by jasonrohrer** [comment]
Yes, this was fixed by ead760bd63dcc2dc36c1edf5059c83cfce369483
